### PR TITLE
[136] Edit existing labels

### PR DESCRIPTION
**Pull request by jbors**
Adds the possibility to click on one of the samples to go back and do some more editing. We also switch from black background JPGs to transparant PNGs for the mask.

It may be a good future improvement to include the mask on the thumbnail to make it more obvious which sample to edit, but out of scope for now.

### CODE REVIEW

**Comment by anna-ruizferrer** [comment]
> Adds the possibility to click on one of the samples to go back and do some more editing. We also switch from black background JPGs to transparant PNGs for the mask.
> 
> It may be a good future improvement to include the mask on the thumbnail to make it more obvious which sample to edit, but out of scope for now.

Everything looks good to me! Not much to comment, because with my null knowledge in kotlin and android it has been more useful to me to read and understand what is going on.

Just a question, for curiosity, the change from JPEG to PNG is done because when setting the mask, at the moment of editing the label, the mask is set over the image, so we need to have it with transparent background?

**Comment by jbors** [comment]
> > Adds the possibility to click on one of the samples to go back and do some more editing. We also switch from black background JPGs to transparant PNGs for the mask.
> > It may be a good future improvement to include the mask on the thumbnail to make it more obvious which sample to edit, but out of scope for now.
> 
> Everything looks good to me! Not much to comment, because with my null knowledge in kotlin and android it has been more useful to me to read and understand what is going on.
> 
> Just a question, for curiosity, the change from JPEG to PNG is done because when setting the mask, at the moment of editing the label, the mask is set over the image, so we need to have it with transparent background?

@anna-ruizferrer Yeah, no such thing as a transparent JPG apparently. Used to be saved as colour on black, now as a transparent mask. Shouldn't make a difference to training according to Edu/Lorenzo

File: app/src/main/java/net/aiscope/gdd_app/ui/mask/customview/MaskLayer.kt [hunk 1]
```diff
@@ -120,10 +126,24 @@ class MaskLayer(private val imageMatrix: Matrix) {
...
+
     private fun pathBeingDrawn() = currentPath != null
 
     private fun composeCurrentStateBitmap() {
         currentStateBitmap.eraseColor(Color.TRANSPARENT)
+
+        initialBitmap?.let{
+            currentStateBitmap = Bitmap.createScaledBitmap(it, size.width, size.height, false)
+            currentStateBitmapCanvas = Canvas(currentStateBitmap)
+        }
```
Comment (eduardb) [inline app/src/main/java/net/aiscope/gdd_app/ui/mask/customview/MaskLayer.kt:145]: Initially I thought that my phone is being slow, but then I realised that it's this that is actually slowing it down. 

This method is ran on each frame, and allocating a bitmap on each frame is really not good for performance. I think that we need to find another way to draw the existing mask that is being edited. I have an idea, but it would involve changing the undo/redo mechanism... Let's catch-up on this topic, if you want.

File: app/src/main/java/net/aiscope/gdd_app/ui/mask/customview/MaskLayer.kt [hunk 2]
```diff
@@ -91,6 +89,14 @@ class MaskLayer(private val imageMatrix: Matrix) {
...
         require(!sizeInitialized()) { "Size was initialized already!" }
         this.size = size
+
+        currentStateBitmap = Bitmap.createBitmap(
+            size.width,
+            size.height,
+            Bitmap.Config.ARGB_8888
+        )
+
+        currentStateBitmapCanvas = Canvas(currentStateBitmap)
```
Comment (eduardb) [inline app/src/main/java/net/aiscope/gdd_app/ui/mask/customview/MaskLayer.kt:99]: Doesn't seem to matter, I think, but I was just curious why did you move these initialisations down here.

File: app/src/main/java/net/aiscope/gdd_app/ui/metadata/MetadataPresenter.kt [hunk 3]
```diff
@@ -74,4 +74,13 @@ class MetadataPresenter @Inject constructor(
         val current = repository.current()
         view.captureImage(current.nextImageName())
     }
+
+    suspend fun editImage(index: Int) {
+        val current = repository.current()
```
Comment (eduardb) [inline app/src/main/java/net/aiscope/gdd_app/ui/metadata/MetadataPresenter.kt:79]: Not a huge deal-breaker, but I think that instead of passing the index around, which is not very reliable (technically, the index in the adapter might be different than the one from the domain model), what if we would pass the mask `File`s down to the adapter along with the image `File`s, which then would make it straightforward to display the mask over the image in the metadata screen (but that part in a future PR maybe). We would kill two birds with one stone.

File: app/src/main/java/net/aiscope/gdd_app/ui/metadata/MetadataActivity.kt [hunk 4]
```diff
@@ -101,6 +104,26 @@ class MetadataActivity : AppCompatActivity(), MetadataView, CaptureFlow {
...
+            setNegativeButton(R.string.capture_flow_exit_dialog_stay) { _, _ ->
+                // do nothing
+            }
+            setMessage(getString(R.string.capture_flow_exit_dialog_message))
+            setTitle(getText(R.string.capture_flow_exit_dialog_title))
+            create()
+        }.show()
+    }
+
+    fun <T> T.goToHome() where T : AppCompatActivity, T : CaptureFlow {
```
[the remark this answers is not in the document]
Comment (jbors) [reply app/src/main/java/net/aiscope/gdd_app/ui/metadata/MetadataActivity.kt:122]: Ah yeah good point 👍

File: app/src/main/java/net/aiscope/gdd_app/ui/metadata/MetadataActivity.kt [hunk 5]
```diff
@@ -101,6 +104,26 @@ class MetadataActivity : AppCompatActivity(), MetadataView, CaptureFlow {
...
+            setPositiveButton(R.string.capture_flow_exit_dialog_exit) { _, _ ->
+                goToHome()
+            }
+            setNegativeButton(R.string.capture_flow_exit_dialog_stay) { _, _ ->
+                // do nothing
+            }
+            setMessage(getString(R.string.capture_flow_exit_dialog_message))
+            setTitle(getText(R.string.capture_flow_exit_dialog_title))
+            create()
+        }.show()
```
[the remark this answers is not in the document]
Comment (jbors) [reply app/src/main/java/net/aiscope/gdd_app/ui/metadata/MetadataActivity.kt]: Sure thing. Done.

File: app/src/main/java/net/aiscope/gdd_app/ui/mask/customview/MaskLayer.kt (hunk 2 above)
Comment (jbors) [reply app/src/main/java/net/aiscope/gdd_app/ui/mask/customview/MaskLayer.kt:99]: @eduardb Well, previously we had a lazily initialized `val currentStateBitmap`, that just happened to be initialized (used) after the `size` var had been set. 

Since we moved it to a `lateinit var` it made sense to me to just initialize the variable when the size gets set.

File: app/src/main/java/net/aiscope/gdd_app/ui/mask/customview/MaskLayer.kt [hunk 6]
```diff
@@ -182,14 +202,7 @@ class MaskLayer(private val imageMatrix: Matrix) {
...
     fun getBitmap(): Bitmap {
-        return Bitmap.createBitmap(
-            currentStateBitmap.width,
-            currentStateBitmap.height,
-            Bitmap.Config.ARGB_8888
-        ).apply {
-            val canvas = Canvas(this)
-            drawPaths(canvas, removeAlpha = true)
-        }
+        return currentStateBitmap.apply { drawPaths(currentStateBitmapCanvas, removeAlpha = true) }
```
[the remark this answers is not in the document]
Comment (jbors) [reply app/src/main/java/net/aiscope/gdd_app/ui/mask/customview/MaskLayer.kt:205]: @eduardb A change is needed to make sure we draw on the `initialBitmap` rather than a new empty one. Think it may be a mistake to use the the `currentStateBitmap` though, will have a look if I can stop using that.

[the remark this answers is not in the document]
Comment (jbors) [reply app/src/main/java/net/aiscope/gdd_app/ui/mask/customview/MaskLayer.kt]: Alright updated implementation. Neater like this I think

File: app/src/main/java/net/aiscope/gdd_app/ui/mask/MaskActivity.kt [hunk 7]
```diff
@@ -167,4 +176,10 @@ class MaskActivity : AppCompatActivity(), MaskView, CaptureFlow {
         options.inPreferredConfig = Bitmap.Config.ARGB_8888
         BitmapFactory.decodeFile(filepath, options)
     }
+
+    private suspend fun readMask(filepath: String): Bitmap? = withContext(Dispatchers.IO) {
+        val bitmap = BitmapFactory.decodeFile(filepath)
+        //Converts to a mutable bitmap
+        bitmap.copy(Bitmap.Config.ARGB_8888, true)
```
[the remark this answers is not in the document]
Comment (jbors) [reply app/src/main/java/net/aiscope/gdd_app/ui/mask/MaskActivity.kt:183]: @eduardb  Well `readImage()` returns a `Bitmap` whereas `readMask()` returns a nullable `Bitmap?` (as it may not exist). I didn't wanna introduce nullability on the normal flow just for the editing so I made `readMask()` into a separate method.

As for moving the part where we make it mutable, sure we could do that in the `MaskLayer` rather than the `MaskActivity` if you prefer.

File: app/src/main/java/net/aiscope/gdd_app/ui/mask/customview/MaskLayer.kt (hunk 1 above)
Comment (jbors) [reply app/src/main/java/net/aiscope/gdd_app/ui/mask/customview/MaskLayer.kt]: Updated as discussed. Thanks!

File: app/src/main/java/net/aiscope/gdd_app/ui/metadata/MetadataPresenter.kt (hunk 3 above)
Comment (jbors) [reply app/src/main/java/net/aiscope/gdd_app/ui/metadata/MetadataPresenter.kt:81]: Alright, updated to use the actual image/mask file

File: app/src/main/java/net/aiscope/gdd_app/ui/metadata/SampleImagesAdapter.kt [hunk 8]
```diff
@@ -80,6 +81,7 @@ private class ImageViewHolder(
                 itemView.context.resources.getDimensionPixelSize(R.dimen.sample_image_thumbnail_height)
             )
             itemView.setImageBitmap(bitmap)
+            itemView.setOnClickListener { onImageClicked(index) }
```
[the remark this answers is not in the document]
Comment (jbors) [reply app/src/main/java/net/aiscope/gdd_app/ui/metadata/SampleImagesAdapter.kt]: @eduardb So this now uses the actual files rather than the index. I still have an `onClickListener` for each bind though. Not entirely sure how to improve this.

[the remark this answers is not in the document]
Comment (jbors) [reply app/src/main/java/net/aiscope/gdd_app/ui/metadata/SampleImagesAdapter.kt]: OK updated. See if this is what you meant
